operator/cluster/cluster.go

```diff
@@ -68,23 +69,9 @@ func AddClusterBase(clientset *kubernetes.Clientset, client *rest.RESTClient, cl
 		return
 	}
 
-	//err = cleanupPreviousTasks(client, cl.Spec.Name, namespace)
-
-	var pvcName string
-
-	_, found, err := kubeapi.GetPVC(clientset, cl.Spec.Name, namespace)
-	if found {
-		log.Debugf("pvc [%s] already present from previous cluster with this same name, will not recreate", cl.Spec.Name)
-		pvcName = cl.Spec.Name
-	} else {
-		pvcName, err = pvc.CreatePVC(clientset, &cl.Spec.PrimaryStorage, cl.Spec.Name, cl.Spec.Name, namespace)
-		if err != nil {
-			log.Error(err)
-			publishClusterCreateFailure(cl, err.Error())
-			return
-		}
-		log.Debugf("created primary pvc [%s]", pvcName)
-	}
+	//get the appropriate PVC name (which is either reused, or just created)
+	//and set that as the primary name
+	primaryName := getOrCreatePVC(clientset, cl, namespace)
```

Comment:
I think this is convoluting creating the PVC and getting the name of the current primary. This should be broken up.

Reply:
This function has been rewritten as follows:

```// verifyOrCreatePVC first determines whether the PVC will use the default cluster name, as is the case
// for a new cluster. If the default name will be used, create this PVC if missing. If the current-primary
// name is not the default, verify that the PVC exists, as this cluster is being recreated after a failover
// or restoration. If this PVC is not present, cluster recreation will necessarily fail.
func verifyOrCreatePVC(clientset *kubernetes.Clientset, cl *crv1.Pgcluster, namespace string) error {
	// if this cluster is using the default PVC name, check whether the PVC exists,
	// and if not, create it for a new cluster
	if isDefaultPVCName(cl) {
		_, pvcExists, err := kubeapi.GetPVC(clientset, cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY], namespace)
		// return error and fail if the error is anything other than a PVC not
		// found error, which is expected when a new cluster is created
		if err != nil && !kerrors.IsNotFound(err) {
			return err
		}
		// if the PVC doesn't exist at this point, it is a new cluster, so
		// a PVC will need to be created
		if !pvcExists {
			if _, err := pvc.CreatePVC(clientset, &cl.Spec.PrimaryStorage, cl.Spec.Name, cl.Spec.Name, namespace); err != nil {
				return err
			}
			log.Debugf("created primary pvc [%s]", cl.Spec.Name)
		}
	} else {
		// check if the PVC exists, as the remaining cases will be when a failover or restoration has occurred.
		// The PVC will have a unique string attached in the form of 'mycluster-abcd' and must already exist for
		// the cluster creation to continue.
		if _, _, err := kubeapi.GetPVC(clientset, cl.Annotations[config.ANNOTATION_CURRENT_PRIMARY], namespace); err != nil {
			return err
		}
	}
	return nil
}```

Because of the _newer_ implementation of 'current primary' tracking in an annotation, the returning of the name is no longer needed. What is still required is the checking for an existing PVC in all cases, and the creation of a new PVC in the case of a new cluster.

Reply:
@tjmoore4 :+1: